pytorch_multi_things_2.py

```diff
@@ -0,0 +1,221 @@
+    opt = context.optimizer(user_created_opt(model))
+
+    # This is not ideal, but I think its not the end of the world.
+    model, opt = context.amp_init(model, opt)
+
+I found out that pytorch 1.6 will have its own amp module, and it's API is quite a bit cleaner than
+the apex.amp module (and it seems to use no gloabls :)
+
+It turns out that the pytorch-native amp API is pretty close to the API I proposed already, so I
+made some minor adjustemnts to make sure that they were even closer (and that we could easily
+support pytorch-native AMP in the future).
+
+The question I still have is: can we automatically handle optimizer.zero_grad()?  Is that too
+strong of an assumption?
+"""
+
+############
+
+"""
+Example 1: pytorch-native gradient scaling.
+
+Notice that there is only one GradScaler in the training loop, just like there is only one
+TrialContext object in a Determined PyTorchTrial.
+
+code from https://pytorch.org/docs/master/notes/amp_examples.html#typical-mixed-precision-training
+"""
+
+# Creates model and optimizer in default precision
+model = Net().cuda()
+optimizer = optim.SGD(model.parameters(), ...)
+
+# Creates a GradScaler once at the beginning of training.
+scaler = GradScaler()
+
+for epoch in epochs:
+    for input, target in data:
+        optimizer.zero_grad()
+
+        # Runs the forward pass with autocasting.
+        with autocast():
+            output = model(input)
+            loss = loss_fn(output, target)
+
+        # Scales loss.  Calls backward() on scaled loss to create scaled gradients.
+        # Backward passes under autocast are not recommended.
+        # Backward ops run in the same precision that autocast used for corresponding forward ops.
+        scaler.scale(loss).backward()
+
+        # scaler.step() first unscales the gradients of the optimizer's assigned params.
+        # If these gradients do not contain infs or NaNs, optimizer.step() is then called,
+        # otherwise, optimizer.step() is skipped.
+        scaler.step(optimizer)
+
+        # Updates the scale for next iteration.
+        scaler.update()
+
+
+# The same thing in Determined:
+
+class MyPytorchTrial(PyTorchTrial):
+    def __init__(self, context):
+        # There's one context, just like there was one GradScaler()
+        # our context object will mimic the API of the GradScaler.
+        self.context = context
+
+        self.scaler = self.context.set_scaler(GradScaler())
+
+        self.model = self.context.model(my_make_model())
+
+        self.optimizer = self.context.optimizer(my_make_optimizer(self.model.parameters()))
+
+    def train_batch(self, batch):
+        input, target = batch
+
+        # I propose we handle zero_grad() automatically... is there a downside to this?
+        # self.optimizer.zero_grad()
+
+        with autocast():
+            output = self.model(input)
+            loss = loss_fn(output, target)
+
+        self.context.scale(loss).backward()
```

Comment:
If we're gonna handle scaler.step() and scaler.update() for them, why not also handle the scale(loss) call inside backward()

Reply:
Shoot, it's been so long I can't remember the pytorch api I was imitating very well.  Off the top of my head, I would say that we should actually *avoid* doing things for the user in the domain of "data science" if the user can do it for themselves.  Whatever Horovod requires of us is sort of the limit to how far we can avoid doing things for the user, since "dtrain" falls squarely on us.

My guess in this case is that the user may have multiple losses and may want to scale some losses but not others.  Admittedly, I never fully understood what @autocast did so I'm not totally sure if it's unambiguous which losses need scaling and which don't.

Also maybe @liamcli or @katport might have some ideas here.